### ISSUE TITLE
try blocks misbehaving on moar backend

### ISSUE THREAD

**Issue by donaldh** [perl6/nqp#170]
Weirdness in try blocks - golfed to this:

> my $ok := 1; try { nqp::die('ohnoes'); $ok := 0; }; nqp::say($ok);

> my $ok := 1; try { nqp::die('ohnoes'); $ok := 0; my $x; }; nqp::say($ok);
1

**Comment by ugexe** [comment]
This appears to be resolved

    perl6 -e 'use nqp; my $ok := 1; try { nqp::die("ohnoes"); $ok := 0; }; nqp::say(nqp::unbox_i($ok));'
    1

**Comment by LemonBoy** [comment]
Compare
```perl6
my $ok := 1; try { nqp::die('ohnoes'); $ok := 42; }; nqp::say($ok);
```
with
```perl6
my $ok := 1; try { nqp::die('ohnoes'); $ok := 42; 0 }; nqp::say($ok);
```

What happens here can be explained by considering that in the former the try block returns `$ok`, while the latter returns a boxed integer. Fast forward to the code-generation phase [1], the `$protil` register now points to either the `$ok` box or the integer one and the point where we land after the exception handler gives back the control [2] to the caller we just stomp over the content of the register that's supposed to hold the `$ok` box by executing `takehandlerresult`, as shown in the following snippet:

```
00021   const_i64_16 loc_5_int, 1
00022   hllboxtype_i loc_7_obj
00023   box_i        loc_7_obj, loc_5_int, loc_7_obj
00024   set          loc_1_obj, loc_7_obj
00025   set          loc_8_obj, loc_9_obj
     annotation: -e:1
00026   getlex       loc_7_obj, lex_Frame_0_$?PACKAGE_obj
00027   decont       loc_10_obj, loc_7_obj
00028   getwho       loc_10_obj, loc_10_obj
00029   const_s      loc_3_str, 'ohnoes'
00030   atkey_o      loc_10_obj, loc_10_obj, loc_3_str
00031   ifnonnull    loc_10_obj, label_4(00042)
00032   wval         loc_12_obj, 0, 0
00033   getwho       loc_12_obj, loc_12_obj
00034   const_s      loc_3_str, 'ohnoes'
00035   atkey_o      loc_12_obj, loc_12_obj, loc_3_str
00036   set          loc_11_obj, loc_12_obj
00037   ifnonnull    loc_12_obj, label_3(00040)
00038   wval         loc_13_obj, 1, 37
00039   set          loc_11_obj, loc_13_obj
     label_3:
00040   set          loc_7_obj, loc_11_obj
00041   goto         label_5(00043)
     label_4:
00042   set          loc_7_obj, loc_10_obj
     label_5:
00043   decont       loc_7_obj, loc_7_obj
00044   smrt_strify  loc_3_str, loc_7_obj
00045   die          loc_7_obj, loc_3_str
     annotation: -e:1
00046   const_i64_16 loc_5_int, 42
00047   hllboxtype_i loc_11_obj
00048   box_i        loc_11_obj, loc_5_int, loc_11_obj
00049   set          loc_1_obj, loc_11_obj
00050   goto         label_6(00052)
00051   takehandlerresult loc_1_obj
     annotation: -e:1
     label_6:
00052   decont       loc_11_obj, loc_1_obj
00053   smrt_strify  loc_3_str, loc_11_obj
00054   say          loc_3_str
00055   null         loc_11_obj
00056   return_o     loc_11_obj
```

I've solved (or work-arounded) this problem by not generating a `takehandlerresult` if the result of the `try` block isn't used at all and it does work, both the nqp and the rakudo spectest runs without hiccups, but I'm not 100% sure it's the right angle to approach this problem.

[1] https://github.com/perl6/nqp/blob/master/src/vm/moar/QAST/QASTOperationsMAST.nqp#L1721
[2] https://github.com/perl6/nqp/blob/master/src/vm/moar/QAST/QASTOperationsMAST.nqp#L1729

**Comment by pmurias** [comment]
It's not the right angle as the bug also happens if the result of the protected block is used.
Introducing a fresh register instead overwriting the one used by the protected block would solve the problem.

**Comment by pmurias** [comment]
Fixed this by fdbebadd1dbffddc470511a04bfb7f859cb276bb.
